Reduce potential for plugin conflict-driven fatal errors during (REST API-based) refund creation

barryhughes opened the pull request:
In the linked issue, we identified the potential for methods to unexpectedly return the wrong type. The specific case we looked at involved the possibility of 'rogue code' hooking into `woocommerce_rest_pre_insert_shop_order_refund_object`, returning an unexpected type and, in doing so, breaking the V2 refunds endpoint.

In this change, we look at introducing a method for handling this sort of problem. *Later,* we can consider rolling it out more generally across other methods.

Closes #47601.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

When testing this, you will need to use a payment gateway that supports automated processing of refunds (COD, BACS, cheque etc **do not** support this). WooPayments, Stripe etc all **do** support this. Alternatively, you could try setting up the linked "Test Cash" gateway, which is based on the cheque gateway, but nominally supports automated refund processing:

🔌 [testcash.zip](https://github.com/user-attachments/files/16036817/testcash.zip)

If you use this, please note you will additionally need to use the **classic** checkout page when placing your test orders (unless you use some other means to create the test orders, like the REST API).

1. Place an order. It will *probably* make your life easier if you purchase a single item. Additionally, it will be a little easier if you disable taxes and shipping (or else purchase an item that is not subject to these extra charges).
2. Via the order editor, identify the order ID, line item ID, and line item cost.
    - To establish the line item ID, you can use your browser's developer tools to inspect the line item. Look for the value of the table row's `data-order_item_id` attribute.
    - If this is too clunky, you can also obtain the information you need via a REST API request: `GET /wp-json/wc/v3/orders/123`.
4. Execute a refund via the REST API (you must use V2). Supposing the order ID is `123` and the line item ID is `456`, and the cost was `10.00`:

```http
POST /wp-json/wc/v2/orders/123/refunds
Authorization: Basic <base64_encoded_key:secret>
Content-Type: application/json

{
  "amount": "10",
  "line_items": [
    {
      "id": "456",
      "refund_total": "10"
    }
  ]
}
```

5. The refund should succeed. 
6. Now add the following snippet, perhaps to a mu-plugin like `wp-content/mu-plugins/test-pr-47918.php`:

```php
add_filter(
	'woocommerce_rest_pre_insert_shop_order_refund_object',
	'__return_false'
);
```

7. Repeat the test with a fresh order.
8. This time, the API request to create the refund should fail, with a payload looking something like this (indicating something went wrong, and that the status of the refund isn't known, so should be verified manually or via a further API request, etc):

```json
{
  "code": "woocommerce_rest_cannot_verify_refund_created",
  "message": "An unexpected error occurred while generating the refund.",
  "data": null
}
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

barryhughes committed:
Reduce potential for plugin conflict-driven fatal errors during (REST API) refund creation.

barryhughes committed:
Parcel up return-type-assurance logic into its own reusable helper.

barryhughes commented:
@coreymckrill this isn't ready for full review yet, but I still wanted to get a provisional review for some thoughts on the basic concept I'm trying out with `Types::ensure_is_type_or_wp_error`.

coreymckrill commented on plugins/woocommerce/src/Internal/Utilities/Types.php:
```diff
@@ -0,0 +1,68 @@
+	 */
+	public static function ensure_is_type_or_wp_error( $thing, string $desired_type, string $message = null, $code = null, bool $data = false ) {
+		// Is the $thing of the expected type?
+		if ( $thing instanceof $desired_type || is_wp_error( $thing ) ) {
```
It looks like `instanceof` only works with objects. If our desired type is e.g. `array` or `integer` I don't think this will work.

barryhughes replied:
> I think we shouldn't assume it will only be used in the return statement at the end of a function, even though that's what's happening in this particular use case.

Agree.

> If our desired type is e.g. array or integer I don't think this will work.

That's right. As currently written, it's not intended to (but the method name probably doesn't make that very clear). 

My initial thought was that, if we're happy with the approach *in general,* we could add different methods like `Types::ensure_is_int_or_false`, to accommodate other common return types and unions. On which note:

> I'm on the fence about whether the or_wp_error part is too opinionated...

It seems like returning `WP_Error|WC_Data` is a very common pattern, in our REST API code in particular. Though, I guess we could have a far more generic and flexible approach that works something like this:

```pseudo
ensure_matches_type( $thing, 'builtin|udt|scalar' )
```

Parsing and evaluating the second param seems feasible, though there might be complexity we don't initially grasp.

barryhughes replied:
Perhaps also worth noting, the utility is in the internal namespace, so we have some freedom and flexibility in terms of being able to modify it in future (in effect, we don't need to care about back-compat for 3PD code).

coreymckrill replied:
> Parsing and evaluating the second param seems feasible, though there might be complexity we don't initially grasp.

There could also be a middle ground. Something like:

```php
ensure_matches_type( $thing, $type, $fallback = null )
```

If `$fallback` is null, enforce the type no matter what. If `$fallback` is valid, use that if the type is wrong. Not sure it would make sense for `$fallback` to be any other arbitrary type, but certainly `WP_Error`, `false`, and maybe even throwing an exception would be good options. That would probably make for some messy logic though. Maybe `$fallback` is a callback, and the alternative return value can be defined there (including the particular error message within the WP_Error/Exception)?

Please excuse my stream of consciousness 😅

barryhughes replied:
That sounds interesting. Let me try something like that and see how it looks.

For the moment, I'll concentrate on objects rather than scalars (we can readily revise this in future, though, because as noted somewhere else the utility is internal).

coreymckrill commented on plugins/woocommerce/src/Internal/Utilities/Types.php:
```diff
@@ -0,0 +1,68 @@
+		$logger = wc_get_logger();
+
+		// Where did the problem come from?
+		$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
```
The logger can generate a backtrace automatically if you add the `backtrace` parameter to the context. E.g.

```php
			$logger->error(
				'Type check failed.',
				array(
					'source'    => 'wc-type-utility',
					'backtrace' => true,
				)
			);
```

barryhughes replied:
I forgot about this, great catch.

coreymckrill reviewed (commented):
* I like this concept as a way to improve type safety within older code where we can't go back and add type hinting without breaking backcompat.
* I think we shouldn't assume it will only be used in the return statement at the end of a function, even though that's what's happening in this particular use case.
* I'm on the fence about whether the `or_wp_error` part is too opinionated...

barryhughes commented:
> where we can't go back and add type hinting without breaking backcompat

We are also constrained by our current minimum PHP requirements, as even for cases where we are defining new methods, or happen to be free to modify a method signature, we cannot yet take advantage of type unions.

barryhughes committed:
Generalize the type checking helper.

barryhughes commented:
@coreymckrill still in ideation mode, hence no corresponding tests, but pushed another change:

- Leave backtrace generation to the logger.
- Generalize the type checking method.
- Implemented your idea of an optional callback to decide what to do in the event of failure.

Interested in your further thoughts here.

barryhughes commented:
[woocommerce_rest_pre_insert_shop_order_refund_object](https://github.com/woocommerce/woocommerce/blob/8.9.1/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-order-refunds-v2-controller.php#L313-L323)

Tangentially related, I'm not sure that `Filters an object before it is inserted via the REST API` is quite accurate. At this point, the refund was potentially already persisted in the db even if a rogue callback changes the return type via the filter hook.

coreymckrill commented:
> I'm not sure that Filters an object before it is inserted via the REST API is quite accurate. At this point, the refund was potentially already persisted in the db even if a rogue callback changes the return type via the filter hook.

That's true, the presence of `wc_create_refund` and `$refund->save_meta_data()` indicates that database operations are already taking place, even though the method is named "prepare_object_for_database". 😕 Not really sure what to do about that...

coreymckrill commented on plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-order-refunds-v2-controller.php:
```diff
@@ -320,7 +322,20 @@ protected function prepare_object_for_database( $request, $creating = false ) {
+		// still need to honor the declared return type.
+		return Types::ensure_instance_of(
+			$refund,
+			WC_Data::class,
```
```suggestion
			WC_Order_Refund::class,
```

This seems pretty generic. Should this be `WC_Order_Refund` instead?

barryhughes replied:
I think that might make sense, certainly that's the datatype we'd expect to see flowing through here in most out-of-the-box scenarios, but if so we should also update the `@return` tag for the method itself, which is currently documented as returning either a `WC_Data` or `WP_Error` object.

We would need some careful consideration before doing that, however. For instance, hypothetically, could there be cases where someone has defined `My_Custom_Refund` as a subclass of `WC_Order` or `WC_Abstract_Order`? If so, narrowing the type to `WC_Order_Refund` in this way would cause breakages, etc.

barryhughes replied:
So, coming back to this: I'm erring on the side of treading lightly, at least initially, and just going with the currently declared return type. However, please flag if you feel that isn't going far enough (I just worry narrowing the acceptable types introduces more risk).

coreymckrill commented on plugins/woocommerce/src/Internal/Utilities/Types.php:
```diff
@@ -0,0 +1,69 @@
+
+		// Invoke the $on_failure handler, if specified.
+		if ( null !== $on_failure ) {
+			return $on_failure( $thing, $desired_type );
```
Would it make sense to include the `$summary` as a parameter for the callback?

barryhughes replied:
I don't *think* so, it's always going to be a very generic explanation and the purpose of the callback is to do its own handling. Not opposed to it, though: do you vote for adding it?

coreymckrill replied:
Nah, it's probably fine. Adding it later would be an additive change, so no harm in leaving it out for now.

coreymckrill commented on plugins/woocommerce/src/Internal/Utilities/Types.php:
```diff
@@ -0,0 +1,69 @@
+		// Summarize the error for use in logging and in case we have to throw an exception.
+		$summary = sprintf(
+			/* translators: %1$s: name of the expected type */
+			__( 'Unable to ensure that an object was of expected type %1$s. This is not necessarily a problem within WooCommerce, and may indicate there is a problem with code from another plugin that has hooked into one or more WooCommerce filters (please refer to the backtrace).', 'woocommerce' ),
```
This probably doesn't need to be translated (see [Logging Best Practices](https://github.com/woocommerce/woocommerce/blob/d2bde1368af18dc3489dcca96086f2986835ee1f/docs/extension-development/logging.md#best-practices)). It also seems pretty verbose to me. Could it maybe be simplified to something like "Object was not the expected type %1$s. Checking the backtrace may reveal where the object was changed."

barryhughes replied:
Yep, we can whittle it down. 

My hope was that giving a more complete steer ('this may be caused by another plugin'), and making it available in the site operator's language, we might ultimately help those folks to self-diagnose and thereby reduce support requests, but perhaps that's a bit naive (and I realize our logging system doesn't support error codes, so sticking to English is probably best, as covered in those best practices).

Will update :+1:

coreymckrill reviewed (commented):
Looking pretty good. I'm curious what @Konamiman thinks about this approach.

barryhughes committed:
Do not replace existing WP_Errors.

Konamiman commented:
> I'm curious what @Konamiman thinks about this approach.
Given that we can't enforce the type of the data returned by hooks, I think such check is a very reasonable thing to do.

barryhughes commented on plugins/woocommerce/src/Internal/Utilities/Types.php:
```diff
@@ -0,0 +1,69 @@
+	 *
+	 * @return mixed
+	 */
+	public static function ensure_instance_of( $thing, string $desired_type, callable $on_failure = null ) {
```
Another thought is, `$thing` could optionally be an array (to support the equivalent of a type union, like `WC_Order|WC_Super_Order`) ... but, that could also be a thing to keep in reserve until it's needed (we have additional latitude to change how this works in future releases, since it is in the internal namespace).

coreymckrill replied:
🤔 Do you mean `$desired_type` could be an array?

barryhughes replied:
Sorry, yes, that's what I meant.

barryhughes replied:
I will not make this change (yet), because:

- [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).
- We can add it later (we are within the internal namespace, so we have that degree of latitude).

barryhughes committed:
Do not translate logging messages, per current best practices.

Ref: https://github.com/woocommerce/woocommerce/blob/d2bde1368af18dc3489dcca96086f2986835ee1f/docs/extension-development/logging.md#best-practices

barryhughes committed:
Add test describing the behaviour of `Types::ensure_instance_of()`.

Note that the exception error code originally specified in the Types class was also removed: if we want this, it needs to be an integer rather than a string. However, we don't have a methodology or centralized registry for handling numeric error codes at present.

barryhughes commented:
> I think such check is a very reasonable thing to do.

So I think we have some consensus that this makes sense, and this feels like a reasonable way to pilot it (applying it to just one controller method initially, and implementing it in the internal namespace so we can pivot or even remove it in future).

I've replied to some questions, applied some suggestions, and added a test describing the expected behavior. Ready for another look :+1:

barryhughes commented:
Sorry ... though it's fine to look at the code, I still need to update the PR description and testing instructions.

barryhughes committed:
Use ternary shorthand, not null coalescing.

barryhughes commented:
I think we're good to go now ... re-requesting review.

barryhughes commented:
There's a test fail in CI that I didn't see locally. It *could* be related, so need to poke further:

```
1) WC_REST_Products_Controller_Tests::test_product_api_get_all_fields
Exception: The SKU (pancake-1) you are trying to insert is already under processing in /var/www/html/wp-content/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:192
```

barryhughes committed:
Add `@since` tag for hook `woocommerce_rest_pre_insert_<POST_TYPE>_object`.

Selected 4.5.0, as this is the point at which the (until then separate) REST API package was merged into WooCommerce Core.

barryhughes committed:
Avoid short ternaries.

barryhughes committed:
Strict types.

coreymckrill reviewed (approved):
👍 Followed the testing instructions with my local env and it all checks out. The entry gets added to a log file and everything.

I was trying to imagine a scenario where this would break backcompat, but since it's already broken in the case where it returns something other than the expected class instance, I don't think that'll be an issue.
